### PR TITLE
Configuration checks

### PR DESCRIPTION
**Pull request by tulinkry**
Approaches #1508 but it's not complete for the urls. However as stated in #1517 it would be nice if the urls did not include the hostname so the checks for it are not possible

For the
 - bugPattern it validates the pattern and if it contains at least one capture group
 - reviewPattern it validates the pattern and if it contains at least one capture group

### CODE REVIEW

File: src/org/opensolaris/opengrok/configuration/Configuration.java [hunk 1]
```diff
@@ -240,15 +242,35 @@ public int getScanningDepth() {
...
+    /**
+     * Set the scanning depth to a new value
+     *
+     * @param scanningDepth the new value
+     * @throws IllegalArgumentException when the scanningDepth is negative
+     */
+    public void setScanningDepth(int scanningDepth) throws IllegalArgumentException {
+        if (scanningDepth < 0) {
+            throw new IllegalArgumentException(
+                    String.format("Invalid value for scanningDepth \"%d\"", scanningDepth));
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: say what is exactly is wrong in the error message

File: src/org/opensolaris/opengrok/configuration/Configuration.java [hunk 2]
```diff
@@ -421,15 +458,35 @@ public int getCachePages() {
...
+    /**
+     * Set the hits per page to a new value
+     *
+     * @param hitsPerPage the new value
+     * @throws IllegalArgumentException when the hitsPerPage is negative
+     */
+    public void setHitsPerPage(int hitsPerPage) throws IllegalArgumentException {
+        if (hitsPerPage < 0) {
+            throw new IllegalArgumentException(
+                    String.format("Invalid value for histPerPage \"%d\"", hitsPerPage));
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: this and all of the above cases that check for negative number could be de-duped. Share the error string, consider new special exception type, etc.

File: src/org/opensolaris/opengrok/configuration/Configuration.java [hunk 3]
```diff
@@ -375,8 +407,13 @@ public int getMessageLimit() {
      * @see RuntimeEnvironment#getMessagesInTheSystem()
      *
      * @param limit the limit
+     * @throws IllegalArgumentException when the limit is negative
      */
-    public void setMessageLimit(int limit) {
+    public void setMessageLimit(int limit) throws IllegalArgumentException {
+        if (limit < 0) {
+            throw new IllegalArgumentException(
+                    String.format("Invalid value for message limit \"%d\"", limit));
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: it should probablty name the attribute, i.e. `messageLimit`

File: src/org/opensolaris/opengrok/configuration/Configuration.java [hunk 4]
```diff
@@ -702,8 +769,18 @@ public String getReviewPattern() {
...
+    /**
+     * Set the review pattern to a new value
+     *
+     * @param reviewPattern the new pattern
+     * @throws PatternSyntaxException when the pattern is not a valid regexp or
+     * does not contain at least one capture group
+     */
+    public void setReviewPattern(String reviewPattern) throws PatternSyntaxException {
+        if (!reviewPattern.matches(".*\\([^\\)]*\\).*")) {
+            throw new PatternSyntaxException("The pattern must contain at least one group.", reviewPattern, 0);
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: similarly to the cases above - the review and bug pattern share error reporting

File: src/org/opensolaris/opengrok/configuration/Configuration.java [hunk 5]
```diff
@@ -682,8 +739,18 @@ public String getBugPage() {
...
-        this.bugPattern = bugPattern;
+    /**
+     * Set the bug pattern to a new value
+     *
+     * @param bugPattern the new pattern
+     * @throws PatternSyntaxException when the pattern is not a valid regexp or
+     * does not contain at least one capture group
+     */
+    public void setBugPattern(String bugPattern) throws PatternSyntaxException {
+        if (!bugPattern.matches(".*\\([^\\)]*\\).*")) {
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: should this be more strict ? i.e. require at least one character inside the parens ?

File: src/org/opensolaris/opengrok/configuration/messages/ConfigMessage.java [hunk 6]
```diff
@@ -182,7 +182,11 @@ protected void invokeSetter(Configuration config, String field, String value) th
                 | IllegalArgumentException ex) {
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/messages/ConfigMessage.java]: what about PatternSyntaxException ?

Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/messages/ConfigMessage.java]: I don't see a case when there is a `PatternSyntaxException`

Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/messages/ConfigMessage.java]: I meant where does the PatternSyntaxException propagate to ?

File: src/org/opensolaris/opengrok/configuration/Configuration.java (hunk 1 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: fixed

File: src/org/opensolaris/opengrok/configuration/Configuration.java (hunk 2 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: shared error string is used for negative numbers

File: src/org/opensolaris/opengrok/configuration/Configuration.java (hunk 3 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: fixed

File: src/org/opensolaris/opengrok/configuration/Configuration.java (hunk 4 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: for the application the empty group is enough but it makes change to restrict the user to have at least a single character in it

File: src/org/opensolaris/opengrok/configuration/Configuration.java (hunk 5 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: for the application the empty group is enough but it makes change to restrict the user to have at least a single character in it

File: src/org/opensolaris/opengrok/configuration/messages/ConfigMessage.java (hunk 6 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/messages/ConfigMessage.java]: added a comment

All exceptions in the invocation are propagated as the `InvocationTargetException` exception

Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/messages/ConfigMessage.java]:
```log
21-Apr-2017 12:44:54.644 WARNING [configurationListener] org.opensolaris.opengrok.configuration.RuntimeEnvironment.handleMessage Message dropped: {0} - message error
 java.io.IOException: Unsupported operation with the configuration for name "hitsPerPage" - Invalid value for "hitsPerPage" - "-5". Expected value greater or equal than 0.
	at org.opensolaris.opengrok.configuration.messages.ConfigMessage.invokeSetter(ConfigMessage.java:202)
	at org.opensolaris.opengrok.configuration.messages.ConfigMessage.applyMessage(ConfigMessage.java:62)
	at org.opensolaris.opengrok.configuration.messages.Message.apply(Message.java:85)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment.handleMessage(RuntimeEnvironment.java:1693)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment$10.run(RuntimeEnvironment.java:1653)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.opensolaris.opengrok.configuration.messages.ConfigMessage.invokeSetter(ConfigMessage.java:165)
	... 5 more
Caused by: java.lang.IllegalArgumentException: Invalid value for "hitsPerPage" - "-5". Expected value greater or equal than 0
	at org.opensolaris.opengrok.configuration.Configuration.setHitsPerPage(Configuration.java:509)
	... 10 more
```

**Pull request merged**